### ISSUE TITLE
[CLOSED] fixes #178

### ISSUE THREAD

**Issue by BlueFinBima** [BlueFinBima/test1#180]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [derammo](https://github.com/derammo)**
_Wednesday Jan 22, 2020 at 12:57 GMT_
_Originally opened as https://github.com/BlueFinBima/Helios/pull/180_

----

This needs to be reviewed.   I decided to make the codes make sense, and not maintain backward compatibility.

Specifically, if someone has {ENTER} (undocumented) or {RETURN} in their profile, they will now get a different result.  It would be pretty obvious as one of their buttons will stop working.    I think we can just release note this in detail instead of going back to something that makes zero sense (RETURN was numpad enter, ENTER was something that isn't a key binding in Windows, and keyboard enter could not be sent at all.

The code was clearly unfinished.

----

_**[derammo](https://github.com/derammo)** included the following code: https://github.com/BlueFinBima/Helios/pull/180/commits_

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Saturday Jan 25, 2020 at 18:19 GMT_

----

@derammo I realise you know all this, but this is just to help me in the future.

From what I can find 0xCA is supposed to be in a reserved range.  I suspect that Gardroc was going to special case it in the same manner as PAUSE (line 207) and didn't get round to it.

*** I'm not sure I understand why PAUSE needs to have an extended scan code.

My interpretation is that the codes in the _keycodes dictionary are the virtual keyboard codes (which don't include a code for NumEnter) see [Windows Virtual Key Codes](https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes). 

I don't think there is a need for Enter_Legacy.  Putting Apple keyboards aside which are named as Windows probably used to be, for us, RETURN=ENTER=0x0D=Main keyboard enter key.  Return should never have caused the enter on the numeric keypad to be pressed.  That is just a bug.   If it were me, I'd probably do what I said Gadroc was trying to do (even though it is wrong and ugly), and make NUMENTER = 0xCA and then force the scancode translation to be 0x049C.  I suspect / hope that a KEY_EXTENDED 0x0d translation will result in a scancode of 0x049C.

What I can't decide is whether we should be doing an extended mapping (MAPVK_VK_TO_VSC_EX) at 206 (in the longer term) and enable the mapping of more keys.

For reference the extended scan codes seem to be
```
#define DIKEYBOARD_ESCAPE 0x0401 //
#define DIKEYBOARD_1 0x0402 //
#define DIKEYBOARD_2 0x0403 //
#define DIKEYBOARD_3 0x0404 //
#define DIKEYBOARD_4 0x0405 //
#define DIKEYBOARD_5 0x0406 //
#define DIKEYBOARD_6 0x0407 //
#define DIKEYBOARD_7 0x0408 //
#define DIKEYBOARD_8 0x0409 //
#define DIKEYBOARD_9 0x040A //
#define DIKEYBOARD_0 0x040B //
#define DIKEYBOARD_MINUS 0x040C // (* - on main keyboard *)
#define DIKEYBOARD_EQUALS 0x040D //
#define DIKEYBOARD_BACK 0x040E // (* backspace *)
#define DIKEYBOARD_TAB 0x040F //
#define DIKEYBOARD_Q 0x0410 //
#define DIKEYBOARD_W 0x0411 //
#define DIKEYBOARD_E 0x0412 //
#define DIKEYBOARD_R 0x0413 //
#define DIKEYBOARD_T 0x0414 //
#define DIKEYBOARD_Y 0x0415 //
#define DIKEYBOARD_U 0x0416 //
#define DIKEYBOARD_I 0x0417 //
#define DIKEYBOARD_O 0x0418 //
#define DIKEYBOARD_P 0x0419 //
#define DIKEYBOARD_LBRACKET 0x041A //
#define DIKEYBOARD_RBRACKET 0x041B //
#define DIKEYBOARD_RETURN 0x041C // (* Enter on main keyboard *)
#define DIKEYBOARD_LCONTROL 0x041D //
#define DIKEYBOARD_A 0x041E //
#define DIKEYBOARD_S 0x041F //
#define DIKEYBOARD_D 0x0420 //
#define DIKEYBOARD_F 0x0421 //
#define DIKEYBOARD_G 0x0422 //
#define DIKEYBOARD_H 0x0423 //
#define DIKEYBOARD_J 0x0424 //
#define DIKEYBOARD_K 0x0425 //
#define DIKEYBOARD_L 0x0426 //
#define DIKEYBOARD_SEMICOLON 0x0427 //
#define DIKEYBOARD_APOSTROPHE 0x0428 //
#define DIKEYBOARD_GRAVE 0x0429 // (* accent grave *)
#define DIKEYBOARD_LSHIFT 0x042A //
#define DIKEYBOARD_BACKSLASH 0x042B //
#define DIKEYBOARD_Z 0x042C //
#define DIKEYBOARD_X 0x042D //
#define DIKEYBOARD_C 0x042E //
#define DIKEYBOARD_V 0x042F //
#define DIKEYBOARD_B 0x0430 //
#define DIKEYBOARD_N 0x0431 //
#define DIKEYBOARD_M 0x0432 //
#define DIKEYBOARD_COMMA 0x0433 //
#define DIKEYBOARD_PERIOD 0x0434 // (* . on main keyboard *)
#define DIKEYBOARD_SLASH 0x0435 // (* / on main keyboard *)
#define DIKEYBOARD_RSHIFT 0x0436 //
#define DIKEYBOARD_MULTIPLY 0x0437 // (* * on numeric keypad *)
#define DIKEYBOARD_LMENU 0x0438 // (* left Alt *)
#define DIKEYBOARD_SPACE 0x0439 //
#define DIKEYBOARD_CAPITAL 0x043A //
#define DIKEYBOARD_F1 0x043B //
#define DIKEYBOARD_F2 0x043C //
#define DIKEYBOARD_F3 0x043D //
#define DIKEYBOARD_F4 0x043E //
#define DIKEYBOARD_F5 0x043F //
#define DIKEYBOARD_F6 0x0440 //
#define DIKEYBOARD_F7 0x0441 //
#define DIKEYBOARD_F8 0x0442 //
#define DIKEYBOARD_F9 0x0443 //
#define DIKEYBOARD_F10 0x0444 //
#define DIKEYBOARD_NUMLOCK 0x0445 //
#define DIKEYBOARD_SCROLL 0x0446 // (* Scroll Lock *)
#define DIKEYBOARD_NUMPAD7 0x0447 //
#define DIKEYBOARD_NUMPAD8 0x0448 //
#define DIKEYBOARD_NUMPAD9 0x0449 //
#define DIKEYBOARD_SUBTRACT 0x044A // (* - on numeric keypad *)
#define DIKEYBOARD_NUMPAD4 0x044B //
#define DIKEYBOARD_NUMPAD5 0x044C //
#define DIKEYBOARD_NUMPAD6 0x044D //
#define DIKEYBOARD_ADD 0x044E // (* + on numeric keypad *)
#define DIKEYBOARD_NUMPAD1 0x044F //
#define DIKEYBOARD_NUMPAD2 0x0450 //
#define DIKEYBOARD_NUMPAD3 0x0451 //
#define DIKEYBOARD_NUMPAD0 0x0452 //
#define DIKEYBOARD_DECIMAL 0x0453 // (* . on numeric keypad *)
#define DIKEYBOARD_OEM_102 0x0456 // (* < > | on UK/Germany keyboards *)
#define DIKEYBOARD_F11 0x0457 //
#define DIKEYBOARD_F12 0x0458 //
#define DIKEYBOARD_F13 0x0464 // (* (NEC PC98) *)
#define DIKEYBOARD_F14 0x0465 // (* (NEC PC98) *)
#define DIKEYBOARD_F15 0x0466 // (* (NEC PC98) *)
#define DIKEYBOARD_KANA 0x0470 // (* (Japanese keyboard) *)
#define DIKEYBOARD_ABNT_C1 0x0473 // (* / ? on Portugese (Brazilian) keyboards *)
#define DIKEYBOARD_CONVERT 0x0479 // (* (Japanese keyboard) *)
#define DIKEYBOARD_NOCONVERT 0x047B // (* (Japanese keyboard) *)
#define DIKEYBOARD_YEN 0x047D // (* (Japanese keyboard) *)
#define DIKEYBOARD_ABNT_C2 0x047E // (* Numpad . on Portugese (Brazilian) keyboards *)
#define DIKEYBOARD_NUMPADEQUALS 0x048D // (* = on numeric keypad (NEC PC98) *)
#define DIKEYBOARD_PREVTRACK 0x0490 // (* Previous Track (DIK_CIRCUMFLEX on Japanese keyboard) *)
#define DIKEYBOARD_AT 0x0491 // (* (NEC PC98) *)
#define DIKEYBOARD_COLON 0x0492 // (* (NEC PC98) *)
#define DIKEYBOARD_UNDERLINE 0x0493 // (* (NEC PC98) *)
#define DIKEYBOARD_KANJI 0x0494 // (* (Japanese keyboard) *)
#define DIKEYBOARD_STOP 0x0495 // (* (NEC PC98) *)
#define DIKEYBOARD_AX 0x0496 // (* (Japan AX) *)
#define DIKEYBOARD_UNLABELED 0x0497 // (* (J3100) *)
#define DIKEYBOARD_NEXTTRACK 0x0499 // (* Next Track *)
#define DIKEYBOARD_NUMPADENTER 0x049C // (* Enter on numeric keypad *)
#define DIKEYBOARD_RCONTROL 0x049D //
#define DIKEYBOARD_MUTE 0x04A0 // (* Mute *)
#define DIKEYBOARD_CALCULATOR 0x04A1 // (* Calculator *)
#define DIKEYBOARD_PLAYPAUSE 0x04A2 // (* Play / Pause *)
#define DIKEYBOARD_MEDIASTOP 0x04A4 // (* Media Stop *)
#define DIKEYBOARD_VOLUMEDOWN 0x04AE // (* Volume - *)
#define DIKEYBOARD_VOLUMEUP 0x04B0 // (* Volume + *)
#define DIKEYBOARD_WEBHOME 0x04B2 // (* Web home *)
#define DIKEYBOARD_NUMPADCOMMA 0x04B3 // (* , on numeric keypad (NEC PC98) *)
#define DIKEYBOARD_DIVIDE 0x04B5 // (* / on numeric keypad *)
#define DIKEYBOARD_SYSRQ 0x04B7 //
#define DIKEYBOARD_RMENU 0x04B8 // (* right Alt *)
#define DIKEYBOARD_PAUSE 0x04C5 // (* Pause *)
#define DIKEYBOARD_HOME 0x04C7 // (* Home on arrow keypad *)
#define DIKEYBOARD_UP 0x04C8 // (* UpArrow on arrow keypad *)
#define DIKEYBOARD_PRIOR 0x04C9 // (* PgUp on arrow keypad *)
#define DIKEYBOARD_LEFT 0x04CB // (* LeftArrow on arrow keypad *)
#define DIKEYBOARD_RIGHT 0x04CD // (* RightArrow on arrow keypad *)
#define DIKEYBOARD_END 0x04CF // (* End on arrow keypad *)
#define DIKEYBOARD_DOWN 0x04D0 // (* DownArrow on arrow keypad *)
#define DIKEYBOARD_NEXT 0x04D1 // (* PgDn on arrow keypad *)
#define DIKEYBOARD_INSERT 0x04D2 // (* Insert on arrow keypad *)
#define DIKEYBOARD_DELETE 0x04D3 // (* Delete on arrow keypad *)
#define DIKEYBOARD_LWIN 0x04DB // (* Left Windows key *)
#define DIKEYBOARD_RWIN 0x04DC // (* Right Windows key *)
#define DIKEYBOARD_APPS 0x04DD // (* AppMenu key *)
#define DIKEYBOARD_POWER 0x04DE // (* System Power *)
#define DIKEYBOARD_SLEEP 0x04DF // (* System Sleep *)
#define DIKEYBOARD_WAKE 0x04E3 // (* System Wake *)
#define DIKEYBOARD_WEBSEARCH 0x04E5 // (* Web Search *)
#define DIKEYBOARD_WEBFAVORITES 0x04E6 // (* Web Favorites *)
#define DIKEYBOARD_WEBREFRESH 0x04E7 // (* Web Refresh *)
#define DIKEYBOARD_WEBSTOP 0x04E8 // (* Web Stop *)
#define DIKEYBOARD_WEBFORWARD 0x04E9 // (* Web Forward *)
#define DIKEYBOARD_WEBBACK 0x04EA // (* Web Back *)
#define DIKEYBOARD_MYCOMPUTER 0x04EB // (* My Computer *)
#define DIKEYBOARD_MAIL 0x04EC // (* Mail *)
#define DIKEYBOARD_MEDIASELECT 0x04ED // (* Media Select *)
```

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Saturday Jan 25, 2020 at 18:25 GMT_

----

BTW the keyboard event tool is a nice asset.  Shame it does not show the scancode and enhanced key code

**Comment by BlueFinBima** [comment]
<a href="https://github.com/derammo"><img src="https://avatars0.githubusercontent.com/u/817160?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [derammo](https://github.com/derammo)**
_Saturday Jan 25, 2020 at 19:40 GMT_

----

the virtual key code for num enter is the same as regular enter (13) and it just sets the source of the event to “numeric keypad” if you set the extended key code bit

**Comment by BlueFinBima** [comment]
<a href="https://github.com/BlueFinBima"><img src="https://avatars0.githubusercontent.com/u/18526232?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [BlueFinBima](https://github.com/BlueFinBima)**
_Sunday Jan 26, 2020 at 14:32 GMT_

----

Fix is candidate for release **after** 1.4.2019.1005
Awesome job @derammo
